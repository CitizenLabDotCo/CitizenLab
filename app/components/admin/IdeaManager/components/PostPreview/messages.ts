import { defineMessages } from 'react-intl';

export default defineMessages({
  edit: {
    id: 'app.components.admin.IdeaManager.PostPreview.edit',
    defaultMessage: 'Edit',
  },
  delete: {
    id: 'app.components.admin.IdeaManager.PostPreview.delete',
    defaultMessage: 'Delete',
  },
  deleteIdeaConfirmation: {
    id: 'app.components.admin.IdeaManager.PostPreview.deleteIdeaConfirmation',
    defaultMessage: 'Are you sure you want to delete this idea?',
  },
  imageAltText: {
    id: 'app.components.admin.IdeaManager.PostPreview.imageAltText',
    defaultMessage: 'Image for idea {ideaTitle}',
  },
  upvoteCount: {
    id: 'app.components.admin.IdeaManager.PostPreview.upvoteCount',
    defaultMessage: 'Upvote count: {count}',
  },
  downvoteCount: {
    id: 'app.components.admin.IdeaManager.PostPreview.downvoteCount',
    defaultMessage: 'Downvote count: {count}',
  },
  voteCounts: {
    id: 'app.components.admin.IdeaManager.PostPreview.voteCounts',
    defaultMessage: 'Vote counts:',
  },
  currentStatus: {
    id: 'app.components.admin.IdeaManager.PostPreview.currentStatus',
    defaultMessage: 'Current status',
  },
  assignee: {
    id: 'app.components.admin.IdeaManager.PostPreview.assignee',
    defaultMessage: 'Assignee',
  },
  openMap: {
    id: 'app.components.admin.IdeaManager.PostPreview.openMap',
    defaultMessage: 'Open Map',
  },
  closeMap: {
    id: 'app.components.admin.IdeaManager.PostPreview.closeMap',
    defaultMessage: 'Close Map',
  },
  picks: {
    id: 'app.components.admin.IdeaManager.PostPreview.picks',
    defaultMessage: 'Picks: {picksNumber}',
  },
  basketsCountTooltip: {
    id: 'app.components.admin.IdeaManager.PostPreview.basketsCountTooltip',
    defaultMessage: 'The number of users that added this idea to their participatory budget',
  },
  cancelEdit: {
    id: 'app.components.admin.IdeaManager.PostPreview.cancelEdit',
    defaultMessage: 'Cancel edit',
  },
  noOne: {
    id: 'app.components.admin.IdeaManager.PostPreview.noOne',
    defaultMessage: 'Unassigned',
  },
  formTitle: {
    id: 'app.components.admin.IdeaManager.PostPreview.formTitle',
    defaultMessage: 'Editing idea',
  },
  save: {
    id: 'app.components.admin.IdeaManager.PostPreview.save',
    defaultMessage: 'Save',
  },
  submitError: {
    id: 'app.components.admin.IdeaManager.PostPreview.submitError',
    defaultMessage: 'Error',
  },
  postedIn: {
    id: 'app.containers.IdeasShow.postedIn',
    defaultMessage: 'Posted in {projectLink}',
  },
});
