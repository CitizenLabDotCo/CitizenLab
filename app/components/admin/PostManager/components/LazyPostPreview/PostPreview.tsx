import React, { lazy, Suspense, PureComponent } from 'react';
import SideModal from 'components/UI/SideModal';
import styled from 'styled-components';
import { colors } from 'utils/styleUtils';
import { ManagerType } from '../..';

// components
import FullPageSpinner from 'components/UI/FullPageSpinner';
import LazyIdeaEdit from './Idea/LazyIdeaEdit';
import LazyIdeaContent from './Idea/LazyIdeaContent';

// lazy-loaded components
const InitiativeContent = lazy(() => import('./Initiative/InitiativeContent'));
const InitiativeEdit = lazy(() => import('./Initiative/InitiativeEdit'));

interface DataProps {}

interface InputProps {
  type: ManagerType;
  onClose: () => void;
  postId: string | null;
  onSwitchPreviewMode: () => void;
  mode: 'edit' | 'view';
}

interface Props extends InputProps, DataProps {}

export const Container = styled.div`
  min-height: 100%;
  width: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Top = styled.div`
  background-color: white;
  border-bottom: 1px solid ${colors.separation};
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  left: 0;
  height: 50px;
  width: 100%;
  padding-left: 10px;
  padding-right: 50px;
  z-index: 1;
`;

export const Content = styled.div`
  padding: 30px;
  margin-top: 50px;
  width: 100%;
`;

export default class PostPreview extends PureComponent<Props> {
  previewComponent = () => {
    const { type, postId, onClose, onSwitchPreviewMode, mode } = this.props;
    const postType = (type === 'AllIdeas' || type === 'ProjectIdeas') ? 'idea' : 'initiative';

    if (postId) {
      return ({
        view: {
          idea: (
            <LazyIdeaContent
              ideaId={postId}
              closePreview={onClose}
              handleClickEdit={onSwitchPreviewMode}
            />
          ),
          initiative: (
            <InitiativeContent
              initiativeId={postId}
              closePreview={onClose}
              handleClickEdit={onSwitchPreviewMode}
            />
          )
        },
        edit: {
          idea: (
            <LazyIdeaEdit
              ideaId={postId}
              goBack={onSwitchPreviewMode}
            />
          ),
          initiative: (
            <InitiativeEdit
              initiativeId={postId}
              goBack={onSwitchPreviewMode}
            />
          )
        }
      }[mode][postType]);
    }

    return null;
  }

  render() {
    const { postId, onClose } = this.props;

    return (
      <SideModal
        opened={!!postId}
        close={onClose}
      >
        <Suspense fallback={<FullPageSpinner />}>
          {this.previewComponent()}
        </Suspense>
      </SideModal>
    );
  }
}
