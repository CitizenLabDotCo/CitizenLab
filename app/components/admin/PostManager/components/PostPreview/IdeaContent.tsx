import React, { PureComponent } from 'react';
import { isNilOrError } from 'utils/helperUtils';
import { adopt } from 'react-adopt';
import { get } from 'lodash-es';

// components
import IdeaAuthor from 'containers/IdeasShow/IdeaAuthor';
import IdeaTitle from 'containers/IdeasShow/IdeaTitle';
import IdeaBody from 'containers/IdeasShow/IdeaBody';
import IdeaMap from 'containers/IdeasShow/IdeaMap';
import OfficialFeedback from 'containers/IdeasShow/OfficialFeedback';
import Comments from 'containers/IdeasShow/Comments';
import FileAttachments from 'components/UI/FileAttachments';
import IdeaSettings from './IdeaSettings';
import VotePreview from './VotePreview';
import InfoTooltip from 'components/admin/InfoTooltip';
import Button from 'components/UI/Button';
import Link from 'utils/cl-router/Link';
import T from 'components/T';
import { Top, Content, Container } from '.';

// srvices
import { deleteIdea } from 'services/ideas';

// resources
import GetResourceFiles, { GetResourceFilesChildProps } from 'resources/GetResourceFiles';
import GetIdea, { GetIdeaChildProps } from 'resources/GetIdea';
import GetIdeaImages, { GetIdeaImagesChildProps } from 'resources/GetIdeaImages';
import GetTenant, { GetTenantChildProps } from 'resources/GetTenant';
import GetProject, { GetProjectChildProps } from 'resources/GetProject';

// i18n
import injectLocalize, { InjectedLocalized } from 'utils/localize';
import { injectIntl, FormattedMessage } from 'utils/cl-intl';
import { InjectedIntlProps, FormattedNumber } from 'react-intl';
import messages from './messages';

// style
import styled from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';
import { darken } from 'polished';

const StyledIdeaTitle = styled(IdeaTitle)`
  margin-bottom: 30px;
`;

const Row = styled.div`
  display: flex;
  width: 100%;
`;

const Left = styled.div`
  flex: 5;
  margin-right: 50px;
  height: 100%;
`;

const BelongsToProject = styled.p`
  width: 100%;
  color: ${colors.label};
  font-weight: 300;
  font-size: ${fontSizes.base}px;
  line-height: normal;
  margin-bottom: 10px;
`;

const ProjectLink = styled(Link)`
  color: inherit;
  font-weight: 400;
  font-size: inherit;
  line-height: inherit;
  text-decoration: underline;
  transition: all 100ms ease-out;
  margin-left: 4px;

  &:hover {
    color: ${darken(0.2, colors.label)};
    text-decoration: underline;
  }
`;

const IdeaImage = styled.img`
  width: 100%;
  margin: 0 0 2rem;
  padding: 0;
  border-radius: 8px;
  border: 1px solid ${colors.separation};
`;

const StyledIdeaBody = styled(IdeaBody)`
  margin-bottom: 20px;
`;

const StyledIdeaMap = styled(IdeaMap)`
  margin-bottom: 40px;
`;

const StyledOfficialFeedback = styled(OfficialFeedback)`
  margin-top: 70px;
`;

const StyledComments = styled(Comments)`
  margin-top: 30px;
`;

const Right = styled.div`
  flex: 2;
  position: sticky;
  top: 80px;
  align-self: flex-start;
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.base}px;
  line-height: 19px;
`;

const BudgetBox = styled.div`
  margin-top: 20px;
  width: 100%;
  height: 95px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 5px;
  position: relative;
  border-radius: 5px;
  background: ${colors.background};
  border: solid 1px ${colors.adminTextColor};
  font-size: ${fontSizes.large}px;
  font-weight: 500;
`;

const Picks = styled.div`
  margin-top: 7px;
  display: flex;
  font-size: ${fontSizes.base}px;
  align-items: center;
`;

interface State {}

interface InputProps {
  ideaId: string | null;
  closePreview: () => void;
  handleClickEdit: () => void;
}

interface DataProps {
  idea: GetIdeaChildProps;
  ideaImages: GetIdeaImagesChildProps;
  ideaFiles: GetResourceFilesChildProps;
  tenant: GetTenantChildProps;
  project: GetProjectChildProps;
}

interface Props extends InputProps, DataProps {}

export class IdeaContent extends PureComponent<Props & InjectedLocalized & InjectedIntlProps, State> {
  handleClickDelete = () => {
    const { idea, closePreview } = this.props;
    const message = this.props.intl.formatMessage(messages.deleteIdeaConfirmation);

    if (!isNilOrError(idea)) {
      if (window.confirm(message)) {
        deleteIdea(idea.id);
        closePreview();
      }
    }
  }

  render() {
    const {
      idea,
      project,
      localize,
      ideaImages,
      ideaFiles,
      tenant,
      handleClickEdit,
      intl: { formatMessage }
    } = this.props;

    if (!isNilOrError(idea)) {
      const ideaTitle = localize(idea.attributes.title_multiloc);
      const ideaImageLarge = !isNilOrError(ideaImages) && ideaImages.length > 0 ? get(ideaImages[0], 'attributes.versions.large', null) : null;
      const ideaGeoPosition = (idea.attributes.location_point_geojson || null);
      const ideaAddress = (idea.attributes.location_description || null);

      return (
        <Container>
          <Top>
            <Button
              icon="edit"
              style="text"
              textColor={colors.adminTextColor}
              onClick={handleClickEdit}
            >
              <FormattedMessage {...messages.edit}/>
            </Button>
            <Button
              icon="delete"
              style="text"
              textColor={colors.adminTextColor}
              onClick={this.handleClickDelete}
            >
              <FormattedMessage {...messages.delete}/>
            </Button>
          </Top>
          <Content>
            {!isNilOrError(project) &&
              <BelongsToProject>
                <FormattedMessage
                  {...messages.postedIn}
                  values={{
                    projectLink:
                      <ProjectLink className="e2e-project-link" to={`/projects/${project.attributes.slug}`}>
                        <T value={project.attributes.title_multiloc} />
                      </ProjectLink>
                  }}
                />
              </BelongsToProject>
            }

            <StyledIdeaTitle
              ideaId={idea.id}
              ideaTitle={ideaTitle}
            />
            <Row>
              <Left>
                {ideaImageLarge &&
                  <IdeaImage src={ideaImageLarge} alt={formatMessage(messages.imageAltText, { ideaTitle })} className="e2e-ideaImage"/>
                }
                <IdeaAuthor
                  authorId={get(idea, 'relationships.author.data.id', null)}
                  ideaCreatedAt={idea.attributes.created_at}
                  ideaId={idea.id}
                />

                <StyledIdeaBody
                  ideaId={idea.id}
                  ideaBody={localize(idea.attributes.body_multiloc)}
                />

                {ideaGeoPosition && ideaAddress &&
                  <StyledIdeaMap
                    address={ideaAddress}
                    position={ideaGeoPosition}
                    id={idea.id}
                  />
                }

                {ideaFiles && !isNilOrError(ideaFiles) &&
                  <FileAttachments files={ideaFiles} />
                }

                <StyledOfficialFeedback
                  ideaId={idea.id}
                />

                <StyledComments ideaId={idea.id} />
              </Left>
              <Right>
                <VotePreview ideaId={idea.id}/>

                {idea.attributes.budget && !isNilOrError(tenant) &&
                  <>
                    <BudgetBox>
                      <FormattedNumber
                        value={idea.attributes.budget}
                        style="currency"
                        currency={tenant.attributes.settings.core.currency}
                        minimumFractionDigits={0}
                        maximumFractionDigits={0}
                      />
                      <Picks>
                        <FormattedMessage {...messages.picks} values={{ picksNumber: idea.attributes.baskets_count }} />
                        &nbsp;
                        <InfoTooltip {...messages.basketsCountTooltip} size="small" position="top-left" />
                      </Picks>
                    </BudgetBox>
                  </>
                }

                <IdeaSettings ideaId={idea.id}/>
              </Right>
            </Row>
          </Content>
        </Container>
      );
    }
    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  idea: ({ ideaId, render }) => <GetIdea id={ideaId}>{render}</GetIdea>,
  project: ({ idea, render }) => <GetProject id={get(idea, 'relationships.project.data.id')}>{render}</GetProject>,
  ideaFiles: ({ ideaId, render }) => <GetResourceFiles resourceId={ideaId} resourceType="idea">{render}</GetResourceFiles>,
  ideaImages: ({ ideaId, render }) => <GetIdeaImages ideaId={ideaId}>{render}</GetIdeaImages>,
  tenant: <GetTenant />,
});

const IdeaContentWithHOCs = injectIntl(injectLocalize(IdeaContent));

const WrappedIdeaContent = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <IdeaContentWithHOCs {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedIdeaContent;
