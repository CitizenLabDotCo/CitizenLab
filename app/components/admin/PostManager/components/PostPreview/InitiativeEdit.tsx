import React from 'react';

// libraries
import { adopt } from 'react-adopt';

// resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetInitiative, { GetInitiativeChildProps } from 'resources/GetInitiative';
import GetInitiativeImages, { GetInitiativeImagesChildProps } from 'resources/GetInitiativeImages';

// utils
import { isNilOrError } from 'utils/helperUtils';

// components
import InitiativesEditFormWrapper from 'containers/InitiativesEditPage/InitiativesEditFormWrapper';
import Button from 'components/UI/Button';
import { Content, Top, Container } from '.';
import FormLocaleSwitcher from 'components/admin/FormLocaleSwitcher';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';
import { colors } from 'utils/styleUtils';

import { Locale } from 'typings';

interface InputProps {
  initiativeId: string;
  goBack: () => void;
}

interface DataProps {
  initiative: GetInitiativeChildProps;
  initiativeImages: GetInitiativeImagesChildProps;
  locale: GetLocaleChildProps;
}

interface Props extends DataProps, InputProps { }

interface State {
  selectedLocale: GetLocaleChildProps;
}

export class InitiativesEditPage extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props as any);
    this.state = {
      selectedLocale: props.locale,
    };
  }

  componentWillReceiveProps(nextProps : Props) {
    const { locale } = nextProps;
    if (!this.state.selectedLocale && this.props.locale !== locale) {
      this.setState({ selectedLocale: locale });
    }
  }

  onLocaleChange = (locale: Locale) => () => {
    this.setState({ selectedLocale: locale });
  }

  render() {
    const { locale, initiative, initiativeImages, goBack } = this.props;
    const { selectedLocale } = this.state;
    if (isNilOrError(locale) || !selectedLocale || isNilOrError(initiative) || initiativeImages === undefined) return null;

    return (
      <Container bgColor={colors.adminBackground}>
        <Top bgColor="white">
          <Button
            icon="arrow-back"
            style="text"
            textColor={colors.adminTextColor}
            onClick={goBack}
          >
            <FormattedMessage {...messages.cancelEdit} />
          </Button>
          <FormLocaleSwitcher
            onLocaleChange={this.onLocaleChange}
            selectedLocale={selectedLocale}
            values={{}}
            noMargin
          />
        </Top>
        <Content>
          <InitiativesEditFormWrapper
            locale={selectedLocale}
            initiative={initiative}
            initiativeImage={isNilOrError(initiativeImages) || initiativeImages.length === 0 ? null : initiativeImages[0]}
            onPublished={goBack}
          />
        </Content>
      </Container>
    );
  }
}

const Data = adopt<DataProps>({
  locale: <GetLocale />,
  initiative: ({ initiativeId, render }) => <GetInitiative id={initiativeId}>{render}</GetInitiative>,
  initiativeImages: ({ initiativeId, render }) => (
    <GetInitiativeImages initiativeId={initiativeId}>{render}</GetInitiativeImages>
  )
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <InitiativesEditPage {...dataProps} {...inputProps} />}
  </Data>
);
