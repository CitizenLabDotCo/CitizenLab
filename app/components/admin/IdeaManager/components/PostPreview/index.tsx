import React, { PureComponent } from 'react';
import SideModal from 'components/UI/SideModal';
import IdeaEdit from './IdeaEdit';
import IdeaContent from './IdeaContent';
import styled from 'styled-components';
import { colors } from 'utils/styleUtils';

interface DataProps {}

interface InputProps {
  onClose: () => void;
  postId: string | null;
  onSwitchPreviewMode: () => void;
  mode: 'edit' | 'view';
}

interface Props extends InputProps, DataProps {}

export const Container = styled.div`
  height: 100%;
  width: 100%;
  position: relative;
`;

export const Top = styled.div`
  display: flex;
  position: fixed;
  top: 0;
  height: 50px;
  width: 100%;
  background-color: ${colors.adminBackground};
  padding-left: 10px;
  padding-right: 50px;
  z-index: 1;
`;

export const Content = styled.div`
  padding: 30px;
  margin-top: 50px;
`;

export default class PostPreview extends PureComponent<Props> {

  render() {
    const { postId, onClose, onSwitchPreviewMode, mode } = this.props;

    return (
      <SideModal
        opened={!!postId}
        close={onClose}
      >
        {mode === 'view' &&
          <IdeaContent
            ideaId={postId}
            closePreview={onClose}
            handleClickEdit={onSwitchPreviewMode}
          />
        }
        {mode === 'edit' && postId &&
          <IdeaEdit
            ideaId={postId}
            goBack={onSwitchPreviewMode}
          />
        }
      </SideModal>
    );
  }
}
