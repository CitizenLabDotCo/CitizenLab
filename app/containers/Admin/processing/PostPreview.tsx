import React, { PureComponent } from 'react';

// components
import IdeaContent from './IdeaContent';
import { Button, Tag } from 'cl2-component-library';

// styling
import styled from 'styled-components';
import { colors, fontSizes, stylingConsts } from 'utils/styleUtils';

// typings
import { ManagerType } from 'components/admin/PostManager';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';
import { addTagging, deleteTagging } from 'services/taggings';
import TagSearch from './TagSearch';
import { isNilOrError } from 'utils/helperUtils';
import injectLocalize, { InjectedLocalized } from 'utils/localize';
import { ITag } from 'services/tags';
import { IMergedTagging } from 'hooks/useTaggings';

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  position: sticky;
  top: ${stylingConsts.menuHeight}px;
  align-items: flex-start;
  height: calc(100vh - ${stylingConsts.menuHeight}px);
  width: 60vw;
  padding: 15px;
  border-left: 1px solid ${colors.adminSeparation};
  > * {
    padding: 0 15px;
  }
`;

const Navigation = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: calc(100vh - ${stylingConsts.menuHeight}px - 30px);
  align-items: center;
  width: 36px;
`;

const StyledNavButton = styled(Button)`
  max-width: 8px;
  margin: 2px;
  button {
    padding: 8px 12px !important;
  }
`;
const TagSection = styled.div`
  height: calc(100vh - ${stylingConsts.menuHeight}px - 30px);
  flex: 2;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  position: sticky;
  top: ${stylingConsts.menuHeight};
  align-self: flex-end;
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.base}px;
  line-height: 19px;
  min-width: 20%;
  border-left: 1px solid ${colors.adminSeparation};
`;

const TagSubSection = styled.div`
  margin: 12px 0px;
  > * {
    padding: 6px 0px;
  }
`;

export const TagList = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 10px;
`;

const StyledTag = styled(Tag)`
  margin: 0px 4px 4px 0px;
  width: fit-content;
`;

interface DataProps {}

type Direction = 'up' | 'down';
interface InputProps {
  type: ManagerType;
  onClose: () => void;
  postId: string | null;
  taggings: IMergedTagging[] | null;
  handleNavigation: (direction: Direction) => void;
  tags: ITag[] | null | undefined;
}

interface Props extends InputProps, DataProps {}

interface State {
  postId: string | null;
  opened: boolean;
  newTag: string | null;
}

class PostPreview extends PureComponent<Props & InjectedLocalized, State> {
  constructor(props: Props & InjectedLocalized) {
    super(props);
    this.state = {
      postId: props.postId,
      opened: false,
      newTag: null,
    };
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.postId !== this.props.postId && this.props.postId) {
      this.setState({ opened: true });
      setTimeout(() => this.setState({ postId: this.props.postId }), 200);
    }
  }

  onClose = () => {
    this.setState({ opened: false });
    this.setState({ postId: null });
    this.props.onClose();
  };

  handleTagInput = (newTag: string) => {
    this.setState({
      newTag,
    });
  };

  handleNavigationClick = (direction: Direction) => (_event) => {
    _event.preventDefault();
    this.props.handleNavigation(direction);
  };

  getManualTaggings = (taggings: IMergedTagging[]) =>
    taggings.filter(
      (tagging) => tagging.attributes.assignment_method === 'manual'
    );
  getAutomaticTaggings = (taggings: IMergedTagging[]) =>
    taggings.filter(
      (tagging) => tagging.attributes.assignment_method === 'automatic'
    );

  tagIdea = (tagId) => () => {
    const postId = this.props.postId;
    postId && addTagging(postId, tagId);
  };

  removeTagging = (taggingId) => () => {
    deleteTagging(taggingId);
  };

  render() {
    const { taggings, localize, tags } = this.props;
    const manualTaggings = isNilOrError(taggings)
      ? []
      : this.getManualTaggings(taggings);
    const automaticTaggings = isNilOrError(taggings)
      ? []
      : this.getAutomaticTaggings(taggings);

    return (
      <Container>
        <Navigation>
          <StyledNavButton
            locale={'en'}
            icon={'close'}
            onClick={this.onClose}
            iconSize={'8px'}
            buttonStyle={'admin-dark-outlined'}
          />
          <div>
            <StyledNavButton
              iconSize={'8px'}
              locale={'en'}
              icon={'chevron-up'}
              buttonStyle={'admin-dark-outlined'}
              onClick={this.handleNavigationClick('up')}
            />
            <StyledNavButton
              iconSize={'8px'}
              locale={'en'}
              icon={'chevron-down'}
              buttonStyle={'admin-dark-outlined'}
              onClick={this.handleNavigationClick('down')}
            />
          </div>
        </Navigation>

        {this.state.postId && (
          <>
            <IdeaContent
              ideaId={this.state.postId}
              manualTaggings={manualTaggings}
            />
            <TagSection>
              {automaticTaggings.length > 0 && (
                <TagSubSection>
                  <FormattedMessage {...messages.addSmartTag} />
                  <TagList>
                    {automaticTaggings.map((tagging) =>
                      tagging.tag ? (
                        <StyledTag
                          key={tagging.id}
                          icon="close"
                          onIconClick={this.removeTagging(tagging.id)}
                          isAutoTag={true}
                          isSelected={false}
                          text={localize(tagging.tag.attributes.title_multiloc)}
                        />
                      ) : null
                    )}
                  </TagList>
                </TagSubSection>
              )}
              {!isNilOrError(tags) && (
                <TagSubSection>
                  <FormattedMessage {...messages.addExistingTag} />
                  <TagList>
                    {tags
                      .filter(
                        (tag) =>
                          !taggings?.find(
                            (tagging) => tagging.attributes.tag_id === tag.id
                          )
                      )
                      .map((tag) => (
                        <StyledTag
                          key={tag.id}
                          onTagClick={this.tagIdea(tag.id)}
                          isAutoTag={false}
                          isSelected={false}
                          text={localize(tag.attributes.title_multiloc)}
                        />
                      ))}
                  </TagList>
                </TagSubSection>
              )}
              <TagSubSection>
                <FormattedMessage {...messages.addNewTag} />
                <TagSearch
                  ideaId={this.state.postId}
                  ideaTagIds={manualTaggings.map((tagging) => tagging.tag?.id)}
                />
              </TagSubSection>
            </TagSection>
          </>
        )}
      </Container>
    );
  }
}

export default injectLocalize(PostPreview);
