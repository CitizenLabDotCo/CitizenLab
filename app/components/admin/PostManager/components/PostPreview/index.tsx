import React, { PureComponent } from 'react';
import SideModal from 'components/UI/SideModal';
import IdeaEdit from './IdeaEdit';
import InitiativeEdit from './InitiativeEdit';
import IdeaContent from './IdeaContent';
import styled from 'styled-components';
import { ManagerType } from '../..';

interface DataProps {}

interface InputProps {
  type: ManagerType;
  onClose: () => void;
  postId: string | null;
  onSwitchPreviewMode: () => void;
  mode: 'edit' | 'view';
}

interface Props extends InputProps, DataProps {}

export const Container = styled.div`
  min-height: 100%;
  width: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Top = styled.div`
  display: flex;
  align-items: center;
  position: fixed;
  top: 0;
  height: 50px;
  width: 100%;
  padding-left: 10px;
  padding-right: 50px;
  z-index: 1;
`;

export const Content = styled.div`
  padding: 30px;
  margin-top: 50px;
`;

export default class PostPreview extends PureComponent<Props> {

  render() {
    const { type, postId, onClose, onSwitchPreviewMode, mode } = this.props;

    return (
      <SideModal
        opened={!!postId}
        close={onClose}
      >
        {mode === 'view' && (type === 'AllIdeas' || type === 'ProjectIdeas') && postId &&
          <IdeaContent
            ideaId={postId}
            closePreview={onClose}
            handleClickEdit={onSwitchPreviewMode}
          />
        }
        {mode === 'edit' && postId && (type === 'AllIdeas' || type === 'ProjectIdeas') && postId &&
          <IdeaEdit
            ideaId={postId}
            goBack={onSwitchPreviewMode}
          />
        }
        {mode === 'edit' && postId && (type === 'Initiatives') &&
          <InitiativeEdit
            initiativeId={postId}
            goBack={onSwitchPreviewMode}
          />
        }
      </SideModal>
    );
  }
}
