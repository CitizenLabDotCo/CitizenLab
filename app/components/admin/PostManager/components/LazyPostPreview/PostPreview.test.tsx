import React from 'react';
import { shallow } from 'enzyme';

jest.mock('./Idea/LazyIdeaEdit', () => 'LazyIdeaEdit');
jest.mock('./Idea/LazyIdeaContent', () => 'LazyIdeaContent');
jest.mock('./Initiative/InitiativeContent', () => 'InitiativeContent');
jest.mock('./Initiative/InitiativeEdit', () => 'InitiativeEdit');

import PostPreview from './PostPreview';

describe('<PostPreview />', () => {
  let closePreview: jest.Mock;
  let onSwitchPreviewMode: jest.Mock;

  beforeEach(() => {
    closePreview = jest.fn();
    onSwitchPreviewMode = jest.fn();
  });

  it('renders correctly closed', () => {
    const Wrapper = shallow(
      <PostPreview
        type="AllIdeas"
        onClose={closePreview}
        postId={null}
        onSwitchPreviewMode={onSwitchPreviewMode}
        mode="view"
      />
    );
    expect(Wrapper).toMatchSnapshot();
    expect(Wrapper.find('SideModal').exists()).toBe(true);
    expect(Wrapper.find('LazyIdeaEdit').exists()).toBe(false);
    expect(Wrapper.find('LazyIdeaContent').exists()).toBe(false);
  });

  it('renders idea content', () => {
    const Wrapper = shallow(
      <PostPreview
        type="ProjectIdeas"
        onClose={closePreview}
        postId="GreatIdea"
        onSwitchPreviewMode={onSwitchPreviewMode}
        mode="view"
      />
    );
    expect(Wrapper).toMatchSnapshot();

    expect(Wrapper.find('SideModal').exists()).toBe(true);
    expect(Wrapper.find('LazyIdeaEdit').exists()).toBe(false);
    expect(Wrapper.find('LazyIdeaContent').exists()).toBe(true);
  });

  it('renders idea edit', () => {
    const Wrapper = shallow(
      <PostPreview
        type="AllIdeas"
        onClose={closePreview}
        postId="GreatIdea"
        onSwitchPreviewMode={onSwitchPreviewMode}
        mode="edit"
      />
    );
    expect(Wrapper).toMatchSnapshot();

    expect(Wrapper.find('SideModal').exists()).toBe(true);
    expect(Wrapper.find('LazyIdeaEdit').exists()).toBe(true);
    expect(Wrapper.find('LazyIdeaContent').exists()).toBe(false);
  });

});
