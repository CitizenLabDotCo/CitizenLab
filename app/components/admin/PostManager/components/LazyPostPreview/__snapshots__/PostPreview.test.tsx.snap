// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostPreview /> renders correctly closed 1`] = `
<SideModal
  close={[MockFunction]}
  opened={false}
>
  <Suspense
    fallback={<FullPageSpinner />}
  />
</SideModal>
`;

exports[`<PostPreview /> renders idea content 1`] = `
<SideModal
  close={[MockFunction]}
  opened={true}
>
  <Suspense
    fallback={<FullPageSpinner />}
  >
    <LazyIdeaContent
      closePreview={[MockFunction]}
      handleClickEdit={[MockFunction]}
      ideaId="GreatIdea"
    />
  </Suspense>
</SideModal>
`;

exports[`<PostPreview /> renders idea edit 1`] = `
<SideModal
  close={[MockFunction]}
  opened={true}
>
  <Suspense
    fallback={<FullPageSpinner />}
  >
    <LazyIdeaEdit
      goBack={[MockFunction]}
      ideaId="GreatIdea"
    />
  </Suspense>
</SideModal>
`;
